process/sync/baseSync.go

```diff
@@ -72,75 +73,104 @@ func (boot *baseBootstrap) loadBlocks(
-		err := boot.store.Has(hdrNonceHashDataUnit, nonceToByteSlice)
-		if err != nil {
-			highestNonceInStorer--
-			break
-		}
-	}
+	highestNonceInStorer := boot.computeHighestNonce(hdrNonceHashDataUnit)
 
 	log.Info(fmt.Sprintf("the highest header nonce committed in storer is %d\n", highestNonceInStorer))
 
-	var err error
-	lastBlocksToSkip := uint64(0)
-
-	for {
-		if highestNonceInStorer <= blockFinality+lastBlocksToSkip {
-			return process.ErrBoostrapFromStorage
-		}
-
-		for i := highestNonceInStorer - blockFinality - lastBlocksToSkip; i <= highestNonceInStorer; i++ {
-			if i > highestNonceInStorer-lastBlocksToSkip {
-				errNotCritical := boot.removeBlock(i, blockUnit, hdrNonceHashDataUnit)
-				if errNotCritical != nil {
-					log.Info(errNotCritical.Error())
-				}
-			} else {
-				err = boot.applyBlock(i, getHeader, getBlockBody)
-				if err != nil {
-					lastBlocksToSkip++
-					break
-				}
+	for currentNonce = highestNonceInStorer; currentNonce > blockFinality; currentNonce-- {
+		for i := currentNonce - blockFinality; i <= currentNonce; i++ {
+			err = boot.applyBlock(i, getHeader, getBlockBody)
+			if err != nil {
+				log.Info(fmt.Sprintf("apply block with nonce %d: %s\n", i, err.Error()))
+				break
 			}
 		}
 
 		if err == nil {
 			err = boot.accounts.RecreateTrie(boot.blkc.GetCurrentBlockHeader().GetRootHash())
 			if err != nil {
-				lastBlocksToSkip++
+				log.Info(fmt.Sprintf("recreate trie for block with nonce %d in shard %d: %s\n",
+					boot.blkc.GetCurrentBlockHeader().GetNonce(),
+					boot.blkc.GetCurrentBlockHeader().GetShardID(),
+					err.Error()))
 				continue
 			}
 
 			break
 		}
 	}
 
+	if currentNonce <= blockFinality {
+		return process.ErrBoostrapFromStorage
+	}
+
+	boostrapRoundIndex, errNotCritical := boot.getBlockRoundFromNonce(currentNonce, getHeader)
+	if errNotCritical != nil {
+		log.Info(fmt.Sprintf(errNotCritical.Error()))
+	} else {
+		boot.boostrapRoundIndex = boostrapRoundIndex
+	}
+
+	for i := currentNonce + 1; i <= highestNonceInStorer; i++ {
+		boot.cleanupStorage(removeBlockBody, i, blockUnit, hdrNonceHashDataUnit)
+	}
+
 	return nil
 }
 
+func (boot *baseBootstrap) computeHighestNonce(hdrNonceHashDataUnit dataRetriever.UnitType) uint64 {
+	highestNonceInStorer := uint64(0)
+
+	for {
+		highestNonceInStorer++
+		nonceToByteSlice := boot.uint64Converter.ToByteSlice(highestNonceInStorer)
+		err := boot.store.Has(hdrNonceHashDataUnit, nonceToByteSlice)
+		if err != nil {
+			highestNonceInStorer--
+			break
+		}
+	}
+
+	return highestNonceInStorer
+}
+
+func (boot *baseBootstrap) cleanupStorage(
+	removeBlockBody func(uint64, dataRetriever.UnitType, dataRetriever.UnitType) error,
+	nonce uint64,
+	blockUnit dataRetriever.UnitType,
+	hdrNonceHashDataUnit dataRetriever.UnitType) {
+
+	errNotCritical := removeBlockBody(nonce, blockUnit, hdrNonceHashDataUnit)
+	if errNotCritical != nil {
+		log.Info(fmt.Sprintf("remove block body with nonce %d: %s\n", nonce, errNotCritical.Error()))
```

Comment:
log.debug ?

Reply:
It can stay like this, as bootstrap from storage is not an frequently event and when we do this we need these info if there would be some problems.